File: src/09.tree/14.102.ts
import { TreeNode } from "./01.100";

export function levelOrder(root: TreeNode | null): number[][] {
    const ans = [];

    if (!root) {
        return ans;
    }

    // BFS
    const queue: TreeNode[] = [root];

    while (queue.length) {
        const temp = [];
        const qq: TreeNode[] = [];
        for (let i = queue.length; i > 0; i--) {
            const node = queue.shift();
            temp.push(node.val);

            if (node.left) {
                qq.push(node.left);
            }
            if (node.right) {
                qq.push(node.right);
            }
        }
        ans.push(temp);
        queue.push(...qq);
    }

    return ans;
}

File: src/09.tree/14.102.go
package tree

func LevelOrder(root *TreeNode) [][]int {
	ans := [][]int{}

	if root == nil {
		return ans
	}

	queue := []*TreeNode{root}

	for len(queue) > 0 {
		temp := []int{}

		for i := len(queue); i > 0; i-- {
			node := queue[0]
			queue = queue[1:]
			temp = append(temp, node.Val)

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		ans = append(ans, temp)
	}

	return ans
}
